### ISSUE TITLE
How is loaded the global state and is there a global sagas ?

### ISSUE THREAD

**Issue by kopax** [mxstbr/react-boilerplate#1277]
I'm having a look at the selectors of App containers and I see : 

    const selectGlobal = () => (state) => state.get('global');

But I can't find where this is loaded into react the 'global' state of the application.

When I was looking on how to add a global states a month ago, I did had to inject my global reducers and global sagas inside the `createRoutes` function.

Is it global just because our landing is App ? Or did you directly load it somewhere ?

I have a logout action and I want it to be stored under a App/sagas.js, is there an equivalent for injecting a globalSagas ?

**Comment by samit4me** [comment]
Is this what your looking for:

```
/**
 * Creates the main reducer with the asynchronously loaded ones
 */
export default function createReducer(asyncReducers) {
  return combineReducers({
    route: routeReducer,
    global: globalReducer,
    language: languageProviderReducer,
    ...asyncReducers,
  });
}
```

https://github.com/mxstbr/react-boilerplate/blob/master/app/reducers.js#L41-L51

**Comment by kopax** [comment]
Thanks! 
Sorry for asking. I don't know why I've looked for `globalReducer` and sublim didn't find this reference.

**Comment by kopax** [comment]
Is there an equivalent for the sagas ?

**Comment by zamrq** [comment]
simplify your life

```javascript
// selector.js
const selectGlobal = () => (state) => state.toJS();

const selectFormContainer = () => createSelector(
  selectFormContainerDomain(),
  selectInitialValues(),
  selectGlobal(),
  (substate, initialValues, global) => Object.assign(substate.toJS(), { initialValues, global })
);
```

const { global } = this.props;
console.log(global) // global state

**Comment by samit4me** [comment]
You are correct, routes.js is where all the action happens! In short, there is no such equivalent. Sagas are different to the reducers.

There can only be a single store and to know the shape of that store you need all the reducers up front, so on route load, if there is a reducer, it is combined with the other reducers and then the entirely new store is generated including existing state and then this new store replaces the existing store.

Sagas, don't require you to have all sagas to be loaded up front as they can be added anytime and sagas can start other sagas. On route load, if there is saga(s) they are also injected in the same [asyncInjectors.js](https://github.com/mxstbr/react-boilerplate/blob/master/app/utils/asyncInjectors.js#L60) file. The function `store.runSaga` will start each saga. This method is setup in [store.js](https://github.com/mxstbr/react-boilerplate/blob/master/app/store.js#L42) on app load. For more info about how that works [check the docs](http://yelouafi.github.io/redux-saga/docs/api/index.html#example).

Hope this helps?

**Comment by kopax** [comment]
@samit4me Yes it really clarified the situation. Thanks!

**Comment by denisflorkin** [comment]
Hi, I just realized I cancel my watcher generator on LOCATION_CHANGE in my "global saga".
Am I correct to think that canceling in the global saga on route change serve no purpose ?

**Comment by RobertSheaO** [comment]
@denisflorkin Can you show an example of how you accomplished a "global saga"? Also, if you've done it how I think, by adding something to the parent route in createRoutes, then yes, canceling on LOCATION_CHANGE would never get fired, as the wrapping "App" container doesn't change location when navigating between childRoutes.

**Comment by denisflorkin** [comment]
Hi @RobertSheaO here is how I set up the global saga is in routes.js :
```JS
export default function createRoutes(store) {
  // Create reusable async injectors using getAsyncInjectors factory
  const { injectReducer, injectSagas } = getAsyncInjectors(store); // eslint-disable-line no-unused-vars

  // inject saga at "root" level (e.g. for no specific routes,
  // it's present on all routes )
  injectSagas(AppWatcher)

  return [
    {
      path: '/',
      name: 'home',
      getComponent(nextState, cb) {
        const importModules = Promise.all([
          System.import('containers/HomePage'),
        ]);

        const renderRoute = loadModule(cb);

        importModules.then(([component]) => {
          renderRoute(component);
        });

        importModules.catch(errorLoading);
      },
    }, {
```

Thanks for the confirmation (I assume I'm indeed doing what you say :
> by adding something to the parent route in createRoutes

**Comment by RobertSheaO** [comment]
@denisflorkin, what is the `AppWatcher` variable? Just an include of your global `sagas.js` file?

**Comment by denisflorkin** [comment]
Yep, the `appWatcher` variable is my "defaultGlobalSaga" (the one that has no route and is active at all time)

**Comment by joelbars** [comment]
I @denisflorkin I know that is been a long time, but how this `AppWatcher` is structured? I tried to make a simple import of my saga file but is not working.

**Comment by justingreenberg** [comment]
this should be resolved by #1746 — feedback is welcome!
